File: Daily Resources/Day02/DSA - Binary Search Answers/Koko Eating Bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        auto can_finish = [&](int speed) {
            long long hours = 0;
            for (int pile : piles) {
                hours += (pile + speed - 1) / speed; // ceil without using floating point
            }
            return hours <= h;
        };

        int left = 1, right = *max_element(piles.begin(), piles.end());
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (can_finish(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
};

File: Daily Resources/Day01/DSA - Binary Search Answers/Koko Eating Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        import math

        def can_finish(speed):
            hours = 0
            for pile in piles:
                hours += math.ceil(pile / speed)
            return hours <= h

        left, right = 1, max(piles)
        while left < right:
            mid = (left + right) // 2
            if can_finish(mid):
                right = mid
            else:
                left = mid + 1
        return left
